lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -71,183 +72,59 @@ class UnusedJavaScript extends ByteEfficiencyAudit {
-    };
-  }
-
   /**
    * @param {LH.Artifacts} artifacts
    * @param {Array<LH.Artifacts.NetworkRequest>} networkRecords
    * @param {LH.Audit.Context} context
    * @return {Promise<ByteEfficiencyAudit.ByteEfficiencyProduct>}
    */
   static async audit_(artifacts, networkRecords, context) {
-    const bundles = artifacts.SourceMaps ? await JSBundles.request(artifacts, context) : [];
+    const bundles = await JsBundles.request(artifacts, context);
     const {bundleSourceUnusedThreshold = IGNORE_BUNDLE_SOURCE_THRESHOLD_IN_BYTES} =
       context.options || {};
 
-    /** @type {Map<string, Array<LH.Crdp.Profiler.ScriptCoverage>>} */
-    const scriptsByUrl = new Map();
-    for (const script of artifacts.JsUsage) {
-      const scripts = scriptsByUrl.get(script.url) || [];
-      scripts.push(script);
-      scriptsByUrl.set(script.url, scripts);
-    }
-
     const items = [];
-    for (const [url, scriptCoverage] of scriptsByUrl.entries()) {
+    for (const [url, scriptCoverages] of Object.entries(artifacts.JsUsage)) {
       const networkRecord = networkRecords.find(record => record.url === url);
       if (!networkRecord) continue;
-      const wasteData = scriptCoverage.map(UnusedJavaScript.computeWaste);
-      const lengths = UnusedJavaScript.determineLengths(wasteData, networkRecord);
       const bundle = bundles.find(b => b.script.src === url);
-      const item = UnusedJavaScript.mergeWaste(wasteData, networkRecord.url, lengths);
-      if (item.wastedBytes <= IGNORE_THRESHOLD_IN_BYTES) continue;
-      if (bundle) {
-        UnusedJavaScript.createBundleMultiData(
-          item, wasteData, bundle, lengths, bundleSourceUnusedThreshold);
+      const unusedJsSummary =
+        await UnusedJavaScriptSummary.request({url, scriptCoverages, bundle}, context);
+      if (unusedJsSummary.wastedBytes <= IGNORE_THRESHOLD_IN_BYTES) continue;
+
+      const transfer = ByteEfficiencyAudit
+        .estimateTransferSize(networkRecord, unusedJsSummary.totalBytes, 'Script');
+      const transferRatio = transfer / unusedJsSummary.totalBytes;
+      const item = {
+        url: unusedJsSummary.url,
+        totalBytes: Math.round(transferRatio * unusedJsSummary.totalBytes),
+        wastedBytes: Math.round(transferRatio * unusedJsSummary.wastedBytes),
+        wastedPercent: unusedJsSummary.wastedPercent,
+      };
+
+      // Augment with bundle data.
+      if (bundle && unusedJsSummary.sourcesWastedBytes) {
+        const topUnusedSourceSizes = Object.entries(unusedJsSummary.sourcesWastedBytes)
+          .slice(0, 5)
+          .map(([source, unused]) => {
+            const total =
+              source === '(unmapped)' ? bundle.sizes.unmappedBytes : bundle.sizes.files[source];
+            return {
+              source,
+              unused: Math.round(unused * transferRatio),
+              total: Math.round(total * transferRatio),
+            };
+          })
+          .filter(d => d.unused >= bundleSourceUnusedThreshold);
```

Comment:
filter moved after the slice and no sort OK because they're all coming from the same network file and `transferRatio` is all the same so  it's a monotonic mapping?

Reply:
I don't *love* the idea of implicitly relying on the sort order of insertion order into the object since it's [not guaranteed](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries), this is small right we could resort just for clarity?

Reply:
1- yup

2- seems like it's de-facto standard by now ... do you envision this ever changing in v8? But we could resort nbd :)

Reply:
> do you envision this ever changing in v8?

not in the time that this line remains untouched :) but more than 50% of my reasoning is for reader clarity more than fear of a current existing bug anyhow. I see things that need to be sorted coming out of a map without a sort and my initial reaction is "how is this not totally wrong?" 😆